### PR TITLE
instance: Add Generation ID support for Virtual Machines

### LINKED ISSUE #11449: Add Generation ID support for Virtual Machines
QEMU supports the concept of a VM Generation ID that is exposed to the guest OS.

This can be used by the hypervisor to indicate to the guest OS that its place in time has moved backwards (such as when an instance is restored from a snapshot or backup).

This allows for the VM guest OS to reinitialize any state it needs to avoid any invalid data that could have occurred since the snapshot or backup was taken (the classic example being replication in a distributed system).

The initial attempt at adding this feature (https://github.com/lxc/lxd/pull/11385) had an issue in that it caused the VM's machine ID to be changed on snapshot and backup restore. This is problematic because it can cause the the guest OS' identity to change, and can also cause it to reconfigure network identifiers.

Instead we need the following changes to support VM generation ID properly:

- [ ] The instance `volatile.uuid` key should remain unchanged during snapshot and backup restores (effectively reverting https://github.com/lxc/lxd/pull/11385). It should also remain unchanged when moved between servers. However it should change when a new instance is created by copying and existing instance (or snapshot). This should be existing behaviour and require no code changes. But worth checking.
- [ ] A new instance key called `volatile.uuid.generation` should be added. This should be set to the same value as `volatile.uuid` at instance creation time (`instance.CreateInternal`) and should also be populated if not populated to the same value as `volatile.uuid` during `qemu.Start()`.
- [ ] Whenever the `volatile.uuid` key is changed the `volatile.uuid.generation` key should be changed to match the `volatile.uuid` value.
- [ ] The `volatile.uuid.generation` value should be passed to QEMU as the VM's generation ID.
- [ ] On snapshot and backup restore a new `volatile.uuid.generation` value should be generated and stored in the restored instance's DB record (the snapshot record should not be modified). And the `volatile.uuid` key should remain unchanged.
- [ ] The daily test for VMs should check that the QEMU process contains the correct Machine ID and VM generation ID flag.
- [x] Depending on the answer to https://github.com/lxc/lxd/issues/11449#issuecomment-1459743902 the test suite can test for the correct population of the `volatile.uuid` and generation UUID keys for containers too.

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes #11449 

- [**READY FOR REVIEW**] The instance volatile.uuid key should remain unchanged during snapshot and backup restores (effectively reverting https://github.com/lxc/lxd/pull/11385). It should also remain unchanged when moved between servers. However it should change when a new instance is created by copying and existing instance (or snapshot). This should be existing behaviour and require no code changes. But worth checking.
- [**READY FOR REVIEW**] A new instance key called `volatile.uuid.vmgenid` should be added. This should be set to the same value as volatile.uuid at instance creation time (instance.CreateInternal) and should also be populated if not populated to the same value as volatile.uuid during qemu.Start().
- [**READY FOR REVIEW**] Whenever the `volatile.uuid` key is changed the `volatile.uuid.vmgenid` key should be changed to match the `volatile.uuid` value.
- [**READY FOR REVIEW**] The `volatile.uuid.vmgenid` value should be passed to QEMU as the VM's generation ID.
- [**READY FOR REVIEW**] On snapshot and backup restore a new `volatile.uuid.vmgenid` value should be generated and stored in the restored instance's DB record (the snapshot record should not be modified). And the `volatile.uuid` key should remain unchanged.
- [**READY FOR REVIEW** => lxc/lxc-ci#713] The daily test for VMs should check that the QEMU process contains the correct Machine ID and VM generation ID flag.
- [**READY FOR REVIEW**] Depending on the answer to https://github.com/lxc/lxd/issues/11449#issuecomment-1459743902 the test suite can test for the correct population of the volatile.uuid and generation UUID keys for containers too.

### CODE REVIEW

File: lxd/instance.go [hunk 1]
```diff
@@ -324,6 +325,7 @@ func instanceCreateAsCopy(s *state.State, opts instanceCreateAsCopyOpts, op *ope
 				// leave alone so we don't prevent copy.
 			}
 
+			srcSnap.LocalConfig()["volatile.uuid"] = uuid.New()
```
Comment (tomponline) [inline lxd/instance.go:328]: The commit message is " instance: volatile.uuid remain unchanged for snapshot/backup restores " and yet this is changing the volatile.uuid, so it suggests it should be in a separate commit (if at all).

File: lxd/instance.go [hunk 2]
```diff
@@ -380,6 +382,8 @@ func instanceCreateAsCopy(s *state.State, opts instanceCreateAsCopyOpts, op *ope
 		}
 	}
 
+	inst.LocalConfig()["volatile.uuid"] = uuid.New()
```
Comment (tomponline) [inline lxd/instance.go:388]: Changing volatile.uuid in a commit that says " instance: volatile.uuid remain unchanged for snapshot/backup restores " doesn't make sense to me im afraid.

**Review by tomponline — COMMENTED** [review]
For the changes to volatile.uuid (and the associated tests) could you put this up as a separate PR that reverts the earlier PR and explains if you are fixing any existing bugs please?

Then the VM genid stuff can be added as a new PR. Thanks

File: lxd/api_internal.go [hunk 3]
```diff
@@ -789,11 +789,11 @@ func internalImportFromBackup(d *Daemon, projectName string, instName string, fo
 		return err
 	}
 
-	// Generate a new UUID for the instance doing a backup.
-	// We will later use this new instance UUID as a VM Generation ID
-	// in order to differentiate the two instances before and after snapshot
-	// restore that changes that instance's place in time (moving it backwards)
-	instDBArgs.Config["volatile.uuid"] = uuid.New()
+	// If the instance is a VM, generate a new `volatile.uuid.vmggenid`
```
Comment (simondeziel) [inline lxd/api_internal.go:792]: Is there an extra "g" in that `vmggenid`? If yes, this was copy-n-pasted to other places.

**Comment by tomponline** [comment]
static analysis failing

File: lxd/api_internal.go [hunk 4]
```diff
@@ -795,6 +795,14 @@ func internalImportFromBackup(d *Daemon, projectName string, instName string, fo
 	// restore that changes that instance's place in time (moving it backwards)
 	instDBArgs.Config["volatile.uuid"] = uuid.New()
 
+	// If the instance is a VM, generate a new `volatile.uuid.vmgenid`
+	// to differentiate this backed up VM from the original.
+	if instDBArgs.Type == instancetype.VM {
+		instDBArgs.Config["volatile.uuid.vmgenid"] = uuid.New()
```
Comment (tomponline) [inline lxd/api_internal.go:801]: My understanding from https://github.com/lxc/lxd/issues/11449 is that we will just use `volatile.uuid.generation` for both VMs and containers. @stgraber do you concur?

**Review by tomponline — CHANGES_REQUESTED** [review]
At least one of the commit messages still references volatile.uuid.vmgenid

File: lxd/api_internal.go [hunk 5]
```diff
@@ -788,6 +789,10 @@ func internalImportFromBackup(d *Daemon, projectName string, instName string, fo
 		return err
 	}
 
+	// If the instance is a VM, generate a new `volatile.uuid.generation`
```
Comment (tomponline) [inline lxd/api_internal.go:792]: The comment says "if the instance is a VM" but this isn't true, we do it for all instance types.

Comment (tomponline) [reply lxd/api_internal.go:792]: I'd expect to see this set within CreateInternal, the same as `volatile.uuid` is.

File: lxd/instance/drivers/driver_qemu.go [hunk 6]
```diff
@@ -1029,6 +1029,13 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 		volatileSet["volatile.uuid"] = instUUID
 	}
 
+	// For a VM instance, we must also set the VM generation ID.
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1032]: This logic is instance type agnostic do I'm surprised to see it in the instance driver. If it has to be here, then its also missing from the lxc driver.

File: lxd/instance/drivers/driver_qemu.go [hunk 7]
```diff
@@ -1284,6 +1291,11 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 		}
 	}
 
+	// VM generation ID only on x86_64 and aarch64.
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1294]: Comment should be:

> // VM generation ID is only used on UEFI compatible architectures.

File: lxd/instance/drivers/driver_qemu.go (hunk 6 above)
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:1032]: This bit is at line 1969 in `driver_lxc.go`

```go
// For a container instance, we must also set the generation UUID.
genUUID := d.localConfig["volatile.uuid.generation"]
if genUUID != "" {
	genUUID = instUUID
	volatileSet["volatile.uuid.generation"] = genUUID
}
```

**Comment by stgraber** [comment]
This PR needs an API extension and doc entry too.

 - For API extension, update doc/api-extensions.md and shared/version/api.go
 - For doc, update doc/reference/instance_options.md

File: lxd/instance.go [hunk 8]
```diff
@@ -324,6 +324,8 @@ func instanceCreateAsCopy(s *state.State, opts instanceCreateAsCopyOpts, op *ope
 				// leave alone so we don't prevent copy.
 			}
 
+			srcSnap.LocalConfig()["volatile.uuid.generation"] = srcSnap.LocalConfig()["volatile.uuid"]
```
Comment (tomponline) [inline lxd/instance.go:327]: This shouldn't be needed because the call to `instance.CreateInternal()` below should populate it to be the same as volatile.uuid.

File: lxd/instance/drivers/driver_qemu.go (hunk 6 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1032]: OK, so no need for the "For <instance type>," bit in the comment, it suggests it only occurs for this specific instance type.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1032]: Also, the logic looks wrong, as it should be `==` not `!=` right?

Suggest something like this:
```go
// Ensure generation UUID is populated.
if d.localConfig["volatile.uuid.generation"] == "" {
	d.localConfig["volatile.uuid.generation"] = instUUID
	volatileSet["volatile.uuid.generation"] = instUUID
}
```

File: lxd/instance/drivers/driver_qemu.go [hunk 9]
```diff
@@ -1284,6 +1291,11 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 		}
 	}
 
+	// VM generation ID is only used on UEFI compatible architectures.
+	if d.architectureSupportsUEFI(d.architecture) {
+		qemuCmd = append(qemuCmd, "-device", fmt.Sprintf("vmgenid,guid=%s", vmGenUUID))
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1296]: Then this can just use `d.localConfig["volatile.uuid.generation"]`

File: lxd/instance/instance_utils.go [hunk 10]
```diff
@@ -768,6 +768,12 @@ func CreateInternal(s *state.State, args db.InstanceArgs, clearLogDir bool) (Ins
 		args.Config["volatile.uuid"] = uuid.New()
 	}
 
+	// Generate a new `volatile.uuid.generation`
+	// to differentiate this backed up instance from the original.
+	if args.Config["volatile.uuid.generation"] == "" {
```
Comment (tomponline) [inline lxd/instance/instance_utils.go:773]: As this is in CreateInternal, is there a particular reason for this logic, rather than just:

```go
args.Config["volatile.uuid.generation"] = args.Config["volatile.uuid"]
```

Comment (tomponline) [reply lxd/instance/instance_utils.go:773]: Also the comment doesn't make sense in all contexts of CreateInternal I'm afraid, so best not have it.

File: lxd/instance_put.go [hunk 11]
```diff
@@ -202,6 +203,10 @@ func instanceSnapRestore(s *state.State, projectName string, name string, snap s
 		}
 	}
 
+	// Generate a new `volatile.uuid.generation`
+	// to differentiate this instance restored from a snapshot from the original instance.
```
Comment (tomponline) [inline lxd/instance_put.go:207]: any reason for the line break here?

File: lxd/instance/drivers/driver_qemu.go (hunk 6 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1032]: The logic here still looks wrong to me? Am i right?

File: lxd/instance/drivers/driver_lxc.go [hunk 12]
```diff
@@ -1966,6 +1966,13 @@ func (d *lxc) startCommon() (string, []func() error, error) {
 		volatileSet["volatile.uuid"] = instUUID
 	}
 
+	// For a container instance, we must also set the generation UUID.
+	genUUID := d.localConfig["volatile.uuid.generation"]
+	if genUUID != "" {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:1971]: This reads to me as "if the instance already has a generation ID then on start, replace it with the instance UUID" which seems wrong to me.

Comment (gabrielmougard) [reply lxd/instance/drivers/driver_lxc.go]: @tomponline I updated it to 
```go
// For a container instance, we must also set the generation UUID.
genUUID := d.localConfig["volatile.uuid.generation"]
if genUUID == "" {
	genUUID = instUUID
	volatileSet["volatile.uuid.generation"] = genUUID
}
```

File: doc/api-extensions.md [hunk 13]
```diff
@@ -2158,3 +2158,9 @@ This adds the following new configuration options for ZFS storage pools:
 * `volume.zfs.block_mode`
 * `volume.block.mount_options`
 * `volume.block.filesystem`
+
+## `instance_generation_id`
+
+Adds support for instance generation ID. The VM or container generation ID will change whenever the instance's place in time moves backwards. This allows for the VM guest OS or container to reinitialize any state it needs to avoid duplicating potential state that has already occurred:
```
Comment (tomponline) [inline doc/api-extensions.md:2164]: We should say here and in the key documentation below, that at this time, the generation ID is only exposed through to VM type instances.

File: test/suites/snapshots.sh [hunk 14]
```diff
@@ -202,8 +203,10 @@ snap_restore() {
   restore_and_compare_fs snap1
 
   # Check that instances UUID remain the same before and after snapshoting
+  # But generation ID should be updated
   newUUID=$(lxc config get bar volatile.uuid)
-  if [ "${initialUUID}" != "${newUUID}" ]; then
+  newGenerationID=$(lxc config get bar volatile.uuid.generation)
+  if [ "${initialGenerationID}" = "${newGenerationID}" ] || [ "${initialUUID}" != "${newUUID}" ]; then
```
Comment (tomponline) [inline test/suites/snapshots.sh:209]: You're using "friendly" errors here, but you've not updated the error message in response to the change in logic.
You should probably have this as a separate check with a separate error message for generation ID.

File: test/suites/snapshots.sh [hunk 15]
```diff
@@ -213,12 +216,14 @@ snap_restore() {
...
     initialUUID=$(lxc config get bar volatile.uuid)
+    initialGenerationID=$(lxc config get bar volatile.uuid.generation)
     lxc start bar
     lxc snapshot bar snap2 --stateful
     restore_and_compare_fs snap2
 
     newUUID=$(lxc config get bar volatile.uuid)
-    if [ "${initialUUID}" != "${newUUID}" ]; then
+    newGenerationID=$(lxc config get bar volatile.uuid.generation)
+    if [ "${initialGenerationID}" = "${newGenerationID}" ] || [ "${initialUUID}" != "${newUUID}" ]; then
```
Comment (tomponline) [inline test/suites/snapshots.sh:226]: Same point about misleading error messages for the logic change.

File: test/suites/snapshots.sh [hunk 16]
```diff
@@ -229,10 +234,12 @@ snap_restore() {
...
   lxc launch testimage bar2
   initialUUID=$(lxc config get bar2 volatile.uuid)
+  initialGenerationID=$(lxc config get bar2 volatile.uuid.generation)
   lxc copy bar2 bar3
   newUUID=$(lxc config get bar3 volatile.uuid)
+  newGenerationID=$(lxc config get bar3 volatile.uuid.generation)
 
-  if [ "${initialUUID}" = "${newUUID}" ]; then
+  if [ "${initialGenerationID}" = "${newGenerationID}" ] || [ "${initialUUID}" = "${newUUID}" ]; then
     echo "==> UUID of the instance should be different after copying snapshot into instance"
```
Comment (tomponline) [inline test/suites/snapshots.sh:243]: UUIDs (plural)

File: doc/api-extensions.md (hunk 13 above)
Comment (gabrielmougard) [reply doc/api-extensions.md:2164]: @tomponline actually that's not true as of now : 
```go
"volatile.uuid.generation": validate.Optional(validate.IsUUID),
```
is part of `InstanceConfigKeysAny`.. Should I put it only in `InstanceConfigKeysVM` then ?

Comment (tomponline) [reply doc/api-extensions.md:2164]: No, leave it as is. You're misunderstanding me.
We are generating the `voliatile.uuid.generation` ID for VMs and containers.
But we are only passing it through into VMs as the VM generation ID.

The documentation should note this important detail.

Comment (gabrielmougard) [reply doc/api-extensions.md:2164]: ah yes I understand

File: doc/api-extensions.md [hunk 17]
```diff
@@ -2158,3 +2158,9 @@ This adds the following new configuration options for ZFS storage pools:
 * `volume.zfs.block_mode`
 * `volume.block.mount_options`
 * `volume.block.filesystem`
+
+## `instance_generation_id`
```
Comment (tomponline) [inline doc/api-extensions.md:2162]: This bit should be in the API commit.

**Review by tomponline — APPROVED** [review]

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Thanks. I'm happy to merge once we get a decent jenkins runs.

**Comment by tomponline** [comment]
Just need to update https://github.com/lxc/lxc-ci/pull/713 now
